p-ata/src/processor.rs

```diff
@@ -0,0 +1,974 @@
+    const SPL_TOKEN_PROGRAM_ID: Pubkey =
+        pinocchio_pubkey::pubkey!("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
+    *program_id == SPL_TOKEN_PROGRAM_ID
+}
+
+/// Check if account data represents an initialized token account.
+/// Mimics p-token's is_initialized_account check.
+///
+/// Panics if account_data.len() < 109.
+#[inline(always)]
+pub(crate) fn is_initialized_account(account_data: &[u8]) -> bool {
+    // Token account state is at offset 108 (after mint, owner, amount, delegate fields)
+    // State: 0 = Uninitialized, 1 = Initialized, 2 = Frozen
+    account_data[108] != 0
+}
+
+/// Validate that account data represents a valid token account.
+/// Replicates Token-2022's GenericTokenAccount::valid_account_data checks.
+#[inline(always)]
+pub(crate) fn valid_token_account_data(account_data: &[u8]) -> bool {
+    // Regular Token account: exact length match and initialized
+    if account_data.len() == TokenAccount::LEN {
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165
+        return is_initialized_account(account_data);
+    }
+
+    // Token-2022's GenericTokenAccount::valid_account_data assumes Multisig
+    // if account_data length is Multisig::LEN. Collisions are prevented by
+    // adding a byte if a token account happens to have the same length as
+    // Multisig::LEN.
+    if account_data.len() > TokenAccount::LEN {
+        if account_data.len() == Multisig::LEN {
+            return false;
+        }
+        // SAFETY: TokenAccount::LEN is compile-ensured to be == 165, and in
+        // this branch account_data.len > TokenAccount::LEN
+        if is_initialized_account(account_data) {
+            return account_data[TokenAccount::LEN] == ACCOUNTTYPE_ACCOUNT;
+        }
+    }
+
+    false
+}
+
+/// Get mint reference from account info
+#[inline(always)]
+pub(crate) fn get_decimals_from_mint(account: &AccountInfo) -> Result<u8, ProgramError> {
+    let mint_data_slice = account.try_borrow_data()?;
+    match unsafe { spl_token_interface::state::load_unchecked::<Mint>(&mint_data_slice) } {
+        Ok(mint) => Ok(mint.decimals),
+        Err(err) => {
+            const MINT_BASE_SIZE: usize = core::mem::size_of::<Mint>();
+            log!(
+                "Error: Mint account data too small. Expected at least {} bytes, found {} bytes",
+                MINT_BASE_SIZE,
+                mint_data_slice.len()
+            );
+            Err(err)
+        }
+    }
+}
+
+/// Get token account reference with validation. Fails if a mutable borrow
+/// of the account has occurred.
+#[inline(always)]
+pub(crate) fn load_token_account(account: &AccountInfo) -> Result<&TokenAccount, ProgramError> {
+    let account_data = account.try_borrow_data()?;
+    if !valid_token_account_data(&account_data) {
+        return Err(ProgramError::InvalidAccountData);
+    }
+    // SAFETY: We've validated the account data structure above
+    unsafe { Ok(&*(account_data.as_ptr() as *const TokenAccount)) }
+}
```

Comment:
Getting the raw pointer (line 164) from the `Ref` (line 159) is problematic. The `Ref` keeps track of borrows, but when this method finished, it will be drop and reset the borrow state. At the same time, the pointer will be live and can lead to UB if there is a mutable borrow after calling this method.

The safe way to implement this is to map the Ref's data reference to a `TokenAccount` reference, so the `Ref` stays alive while the reference is alive.

```suggestion
pub(crate) fn load_token_account(account: &AccountInfo) -> Result<Ref<TokenAccount>, ProgramError> {
    let account_data = account.try_borrow_data()?;
    if !valid_token_account_data(&account_data) {
        return Err(ProgramError::InvalidAccountData);
    }
    // SAFETY: We've validated the account data structure above
    Ok(Ref::map(account_info.try_borrow_data()?, |data| unsafe {
        &*(data.as_ptr() as *const TokenAccount)
    })
}
```